keps/sig-network/0007-ipvs-proxier.md

```diff
@@ -0,0 +1,571 @@
+
+```shell
+# ipvsadm -ln
+IP Virtual Server version 1.2.1 (size=4096)
+Prot LocalAddress:Port Scheduler Flags
+  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn     
+TCP  10.102.128.4:3080 rr
+  -> 10.244.0.235:8080            Masq    1      0          0         
+  -> 10.244.1.237:8080            Masq    1      0          0     
+
+```
+
+#### Falling back to iptables
+
+IPVS proxier will employ iptables in doing packet filtering, SNAT and supporting NodePort type service. Specifically, ipvs proxier will fall back on iptables in the following 4 scenarios.
+
+* kube-proxy start with --masquerade-all=true
+* Specify cluster CIDR in kube-proxy startup
+* Load Balancer Source Ranges is specified for LB type service
+* Support NodePort type service
+
+And, IPVS proxier will maintain 5 kubernetes-specific chains in nat table
+
+- KUBE-POSTROUTING
+- KUBE-MARK-MASQ
+- KUBE-MARK-DROP
+- KUBE-MARK-SERVICES
+- KUBE-NODEPORTS
+
+`KUBE-POSTROUTING`, `KUBE-MARK-MASQ`, ` KUBE-MARK-DROP` are maintained by kubelet and ipvs proxier won't create them. IPVS proxier will make sure chains `KUBE-MARK-SERVICES` and `KUBE-NODEPORTS` exist in its sync loop.
+
+**1. kube-proxy start with --masquerade-all=true**
+
+If kube-proxy starts with `--masquerade-all=true`, the IPVS proxier will masquerade all traffic accessing service ClusterIP, which behaves same as what iptables proxier does.
+Suppose there is a serivice with Cluster IP `10.244.5.1` and port `8080`:
+
+```shell
+# iptables -t nat -nL
+
+Chain PREROUTING (policy ACCEPT)
+target     prot opt source               destination         
+KUBE-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes service portals */
+
+Chain OUTPUT (policy ACCEPT)
+target     prot opt source               destination         
+KUBE-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes service portals */
+
+Chain POSTROUTING (policy ACCEPT)
+target     prot opt source               destination         
+KUBE-POSTROUTING  all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes postrouting rules */
+
+Chain KUBE-POSTROUTING (1 references)
+target     prot opt source               destination         
+MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes service traffic requiring SNAT */ mark match 0x4000/0x4000
+
+Chain KUBE-MARK-DROP (0 references)
+target     prot opt source               destination         
+MARK       all  --  0.0.0.0/0            0.0.0.0/0            MARK or 0x8000
+
+Chain KUBE-MARK-MASQ (6 references)
+target     prot opt source               destination         
+MARK       all  --  0.0.0.0/0            0.0.0.0/0            MARK or 0x4000
+
+Chain KUBE-SERVICES (2 references)
+target     prot opt source               destination         
+KUBE-MARK-MASQ  tcp  -- 0.0.0.0/0        10.244.5.1            /* default/foo:http cluster IP */ tcp dpt:8080
+```
+
+**2. Specify cluster CIDR in kube-proxy startup**
```

Comment:
The kube-proxy also writes FORWARD chain rules in this scenario as seen here: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L1247-L1268

We might need to implement that behavior as well here?

Reply:
Can we add it in a followup PR?